### ISSUE TITLE
Demo `ipytone` with JupyterLite

### ISSUE THREAD

**Issue by jtpio** [benbovy/ipytone#61]
`ipytone` is a cool project!

Looks like it should also run in JupyterLite since it's a widget?

In that case this could be a way to easily demo the library without requesting a new server on mybinder.org. For example by using `jupyterlite-sphinx` so the JupyterLite website is automatically deployed to ReadTheDocs: https://jupyterlite-sphinx.readthedocs.io/en/latest/

**Comment by benbovy** [comment]
Thanks, and yes that would be great indeed!

I did try a while ago integrating ipytone in Jupyterlite and it worked well.

I haven't spend much time on writing the documentation yet (I'd like to first add a few more Tone.js features like events, instruments, etc.), but I'll use Jupyterlite as soon as the examples are ready!

Thanks for your work on all these wonderful tools!

**Comment by benbovy** [comment]
@jtpio I'm working on the documentation in #72 and enabled jupyterlite with jupyterlite-sphinx. It seems to work well although I have two questions:

- should we be able to see jupyterlite when building and opening the docs locally? Jupyterlite (the `jupyterlite` directive) is shown correctly on RTD (via the PR build), but this is what I see with the local build:

<img width="784" alt="Screenshot 2022-05-30 at 15 05 21" src="https://user-images.githubusercontent.com/4160723/170999431-cef0212a-fa05-40c1-97ce-6906f6232e86.png">

- I'm using https://xeus-python-kernel.readthedocs.io as it looks like it is easier to include the packages that we need. However, I'm wondering it supports a development installation of `ipytone` (so that the docs can use it). Or does it only work only with published pip or conda packages? If it's supported, is there any extra step that I'm missing? I've tried `import ipytone` from jupyterlite on RTD but it's not working.

**Comment by jtpio** [comment]
Looking great @benbovy!

> should we be able to see jupyterlite when building and opening the docs locally? 

I would expect it to work locally as well. Is the docs site served with an HTTP server, for example `python -m http.server`?

> I've tried to import ipytone from jupyterlite on RTD but it's not working.

The xeus python kernel approach should be able to handle noarch packages like widgets normally. cc @martinRenou who might know more about it.

**Comment by martinRenou** [comment]
Did you install it with piplite prior to trying to import it?

```python
# Cell 1
import piplite
await piplite.install("ipytone")

# Cell 2
import ipytone
```

**Comment by martinRenou** [comment]
> I'm using https://xeus-python-kernel.readthedocs.io/ as it looks like it is easier to include the packages that we need

Nice to see you're using it!

> However, I'm wondering if it supports a development installation of ipytone (so that the docs can use it)

I haven't put much thought into it, but using a development installation will be difficult for sure. It requires you to make a local conda channel for emscripten-32 platform where you've built ipytone with boa and making some symlinks here and there.

> Or does it only work only with a stable version published via pip or conda? If it's supported, is there any extra step that I'm missing?

You should just need to install it with the CLI option that xeus-python-kernel adds to `jupyter lite build`. 

If that is not enough, are you using a local build? I will bet on cache issues. Please make sure to clean all browser data for your page.

**Comment by martinRenou** [comment]
You might want to check https://github.com/martinRenou/ipycanvas/pull/275.

**Comment by benbovy** [comment]
Thanks @jtpio and @martinRenou!

> Is the docs site served with an HTTP server, for example python -m http.server?

Ah yes, I did open `index.html` directly in my browser. With the server it works now!

> You might want to check https://github.com/martinRenou/ipycanvas/pull/275.

Yes I shamelessly copied parts of the ipycanvas docs config files in #72. However, I'm still unable to import 3rd party packages in Jupyterlite (ipytone, ipywidgets, numpy, etc.), though. Not sure what's going on. Here is the output of the JupyterLite build called during the sphinx build:


<details>
<summary>**JupyterLite build log**</summary>

```
[jupyterlite-sphinx] Running JupyterLite build
[LiteBuildApp] Searching ['/Users/me/ipytone/docs', '/Users/me/.jupyter', '/Users/me/.local/etc/jupyter', '/Users/me/miniconda3/envs/ipytone_dev/etc/jupyter', '/usr/local/etc/jupyter', '/etc/jupyter'] for config files
[LiteBuildApp] Looking for jupyter_config in /etc/jupyter
[LiteBuildApp] Looking for jupyter_config in /usr/local/etc/jupyter
[LiteBuildApp] Looking for jupyter_config in /Users/me/miniconda3/envs/ipytone_dev/etc/jupyter
[LiteBuildApp] Looking for jupyter_config in /Users/me/.local/etc/jupyter
[LiteBuildApp] Looking for jupyter_config in /Users/me/.jupyter
[LiteBuildApp] Looking for jupyter_config in /Users/me/ipytone/docs
[LiteBuildApp] Looking for jupyterlite_config in /Users/me/ipytone/docs
[LiteBuildApp] [lite] [addon] loading ...
[LiteBuildApp] [lite] [addon] [jupyterlite-xeus-python] load ...
[LiteBuildApp] [lite] [addon] [jupyterlite-xeus-python] ... will post_build
[LiteBuildApp] [lite] [addon] [jupyterlite-xeus-python] ... will pre_build
[LiteBuildApp] [lite] [addon] [archive] load ...
[LiteBuildApp] [lite] [addon] [archive] ... will archive
[LiteBuildApp] [lite] [addon] [archive] ... will status
[LiteBuildApp] [lite] [addon] [contents] load ...
[LiteBuildApp] [lite] [addon] [contents] ... will build
[LiteBuildApp] [lite] [addon] [contents] ... will check
[LiteBuildApp] [lite] [addon] [contents] ... will post_build
[LiteBuildApp] [lite] [addon] [contents] ... will status
[LiteBuildApp] [lite] [addon] [federated_extensions] load ...
[LiteBuildApp] [lite] [addon] [federated_extensions] ... will post_build
[LiteBuildApp] [lite] [addon] [federated_extensions] ... will post_init
[LiteBuildApp] [lite] [addon] [federated_extensions] ... will pre_build
[LiteBuildApp] [lite] [addon] [lite] load ...
[LiteBuildApp] [lite] [addon] [lite] ... will build
[LiteBuildApp] [lite] [addon] [lite] ... will check
[LiteBuildApp] [lite] [addon] [lite] ... will status
[LiteBuildApp] [lite] [addon] [mathjax] load ...
[LiteBuildApp] [lite] [addon] [mathjax] ... will build
[LiteBuildApp] [lite] [addon] [mathjax] ... will check
[LiteBuildApp] [lite] [addon] [mathjax] ... will post_build
[LiteBuildApp] [lite] [addon] [mathjax] ... will status
[LiteBuildApp] [lite] [addon] [piplite] load ...
[LiteBuildApp] [lite] [addon] [piplite] ... will build
[LiteBuildApp] [lite] [addon] [piplite] ... will check
[LiteBuildApp] [lite] [addon] [piplite] ... will post_build
[LiteBuildApp] [lite] [addon] [piplite] ... will post_init
[LiteBuildApp] [lite] [addon] [pyodide] load ...
[LiteBuildApp] [lite] [addon] [pyodide] ... will build
[LiteBuildApp] [lite] [addon] [pyodide] ... will check
[LiteBuildApp] [lite] [addon] [pyodide] ... will post_build
[LiteBuildApp] [lite] [addon] [pyodide] ... will post_init
[LiteBuildApp] [lite] [addon] [pyodide] ... will status
[LiteBuildApp] [lite] [addon] [report] load ...
[LiteBuildApp] [lite] [addon] [report] ... will pre_archive
[LiteBuildApp] [lite] [addon] [serve] load ...
[LiteBuildApp] [lite] [addon] [serve] ... will serve
[LiteBuildApp] [lite] [addon] [serve] ... will status
[LiteBuildApp] [lite] [addon] [settings] load ...
[LiteBuildApp] [lite] [addon] [settings] ... will check
[LiteBuildApp] [lite] [addon] [settings] ... will post_build
[LiteBuildApp] [lite] [addon] [settings] ... will pre_build
[LiteBuildApp] [lite] [addon] [settings] ... will status
[LiteBuildApp] [lite] [addon] [static] load ...
[LiteBuildApp] [lite] [addon] [static] ... will init
[LiteBuildApp] [lite] [addon] [static] ... will post_init
[LiteBuildApp] [lite] [addon] [static] ... will pre_init
[LiteBuildApp] [lite] [addon] [static] ... will pre_status
[LiteBuildApp] [lite] [addon] [translation] load ...
[LiteBuildApp] [lite] [addon] [translation] ... will build
[LiteBuildApp] [lite] [addon] [translation] ... will check
[LiteBuildApp] [lite] [addon] [translation] ... will status
[LiteBuildApp] [lite] [addon] ... OK 13 addons
[LiteBuildApp] [lite] [tasks] loading ...
[LiteBuildApp] [lite] [tasks] ... OK 18 tasks
[LiteBuildApp] No archive (yet): tmpsj3g6lqd-jupyterlite.tgz
[LiteBuildApp] [lite] [contents] All Contents ['/Users/me/ipytone/docs/source/_contents/basic_example_nb.ipynb']
[LiteBuildApp]     jupyter-lite.(json|ipynb): []
[LiteBuildApp]     jupyter-lite.(json|ipynb): 0 files
[LiteBuildApp] [lite] [translation] All Translations []
[LiteBuildApp] creating folder /Users/me/ipytone/docs/build/html/lite/extensions
[LiteBuildApp] creating folder /Users/me/ipytone/docs/build/html/lite/extensions/@jupyter-widgets
[LiteBuildApp] creating folder /Users/me/ipytone/docs/build/html/lite/extensions/@jupyterlite
[LiteBuildApp] creating folder /Users/me/ipytone/docs/build/html/lite/files
[jupyterlite-sphinx] JupyterLite build done
```

</details>


> I haven't put much thought into it, but using a development installation will be difficult for sure.

Ah yes, it looks quite difficult indeed!

I'm thinking about what would be best for the user guide. Using Jupyterlite for the whole user guide would be great, actually. Statically rendered notebooks don't really help here since `ipytone` widgets have no visual output, and showing un-executed code snippets without any output would not be a great user experience either. On the other hand, I'm concerned about how to incrementally update the user guide if the new features cannot be used through Jupyterlite. Any idea about a possible workaround or alternative solution?

**Comment by benbovy** [comment]
@jtpio any chance of making a release of JupyterLite soon? The jupyterlite build in #72 is working locally with the recent fixes in https://github.com/jupyterlite/jupyterlite/pull/644.

**Comment by jtpio** [comment]
@benbovy yes I was thinking about getting https://github.com/jupyterlite/jupyterlite/pull/641 in and making a new release right after.

**Comment by jtpio** [comment]
@benbovy FYI the new release is out: https://github.com/jupyterlite/jupyterlite/releases/tag/v0.1.0b8

**Comment by benbovy** [comment]
Thanks @jtpio! It is working great! https://ipytone--72.org.readthedocs.build/en/72/demo.html

**Comment by davidbrochart** [comment]
This ~~looks~~ sounds great @benbovy, awesome work on ipytone!

**Comment by benbovy** [comment]
Thanks @davidbrochart!

**Comment by benbovy** [comment]
Closed in #72.

**Comment by jtpio** [comment]
Nice!
